File: packages/job-search/src/coach.ts
import type {
  CareerCoachSessionDto,
  CareerGrowthHubDto,
  CoachContextDto,
  CoachFocus,
  CoachMessageDto,
} from '@jobmatch/types';
import { COACH_FOCUS_LABELS, isCoachFocus } from '@jobmatch/types';

export type CoachReply = {
  content: string;
  source: 'template' | 'llm';
};

export function snapCoachContext(
  hub: CareerGrowthHubDto,
  memory?: { summary?: string | null; facts?: string[] } | null,
): CoachContextDto {
  return {
    summary: hub.summary,
    topGaps: hub.skillGaps.slice(0, 6).map((g) => ({
      skill: g.skill,
      priority: g.priority,
      reason: g.reason,
    })),
    roadmapSteps: hub.roadmap.slice(0, 6).map((s) => ({
      title: s.title,
      skill: s.skill,
      estimatedHours: s.estimatedHours,
    })),
    certifications: hub.certifications.slice(0, 4).map((c) => ({
      name: c.name,
      skill: c.skill,
    })),
    careerPaths: hub.careerPaths.slice(0, 3).map((p) => ({
      title: p.title,
      readinessPct: p.readinessPct,
      detail: p.detail,
    })),
    salaryDetail: hub.salaryGrowth?.detail ?? null,
    promotion: {
      score: hub.promotionReadiness.score,
      level: hub.promotionReadiness.level,
      targetSeniority: hub.promotionReadiness.targetSeniority,
      detail: hub.promotionReadiness.detail,
      checklistOpen: hub.promotionReadiness.checklist
        .filter((item) => !item.done)
        .slice(0, 4)
        .map((item) => item.label),
    },
    market: hub.market,
    memorySummary: memory?.summary?.trim() || '',
    memoryFacts: memory?.facts?.slice(0, 12) ?? [],
  };
}

export function sessionTitle(focus: CoachFocus, firstUserMessage?: string | null): string {
  const label = COACH_FOCUS_LABELS[focus];
  const trimmed = firstUserMessage?.trim();
  if (trimmed && trimmed.length > 0) {
    return trimmed.length > 64 ? `${trimmed.slice(0, 61)}…` : trimmed;
  }
  return `${label} session`;
}

export function normalizeFocus(value?: string | null): CoachFocus {
  if (value && isCoachFocus(value)) return value;
  return 'general';
}

export function newMessageId(): string {
  return `msg_${Date.now().toString(36)}_${Math.random().toString(36).slice(2, 8)}`;
}

export function makeMessage(
  role: CoachMessageDto['role'],
  content: string,
  source?: CoachMessageDto['source'],
): CoachMessageDto {
  return {
    id: newMessageId(),
    role,
    content: content.trim(),
    ...(role === 'assistant' && source ? { source } : {}),
    createdAt: new Date().toISOString(),
  };
}

export function buildWelcomeReply(context: CoachContextDto, focus: CoachFocus): CoachReply {
  const gaps = context.topGaps.map((g) => g.skill).slice(0, 3);
  const gapLine = gaps.length
    ? `Your top skill gaps right now are ${gaps.join(', ')}.`
    : 'Your profile already covers most high-demand skills in the catalog.';

  const focusOpeners: Record<CoachFocus, string> = {
    skill_gaps: `Let's close the skills that matter most for roles you're targeting. ${gapLine}`,
    roadmap: context.roadmapSteps[0]
      ? `Your Growth Hub roadmap starts with “${context.roadmapSteps[0].title}”. I can help you sequence the next steps.`
      : 'Your roadmap is thin until more market gaps show up — tell me what role you want next.',
    salary: context.salaryDetail
      ? `On compensation: ${context.salaryDetail}`
      : 'I can help you frame salary expectations once more catalog roles match your currency and level.',
    promotion: `Promotion readiness is ${context.promotion.score}% toward ${context.promotion.targetSeniority}. ${context.promotion.detail}`,
    career_path: context.careerPaths[0]
      ? `A strong path on your hub is “${context.careerPaths[0].title}” (${context.careerPaths[0].readinessPct}% ready).`
      : 'Add desired roles on your profile so I can suggest clearer career ladders.',
    general: `I'm your career coach grounded in the Growth Hub (not generic advice). ${gapLine} Ask about skills, roadmap, salary, or promotion readiness.`,
  };

  const next = nextActions(context, focus);
  return {
    source: 'template',
    content: `${focusOpeners[focus]}\n\n${next}`,
  };
}

export function buildTemplateCoachReply(input: {
  context: CoachContextDto;
  focus: CoachFocus;
  userMessage: string;
  history: CoachMessageDto[];
}): CoachReply {
  const text = input.userMessage.toLowerCase();
  const inferred = inferFocusFromMessage(text, input.focus);

  let body: string;
  switch (inferred) {
    case 'salary':
      body = input.context.salaryDetail
        ? `${input.context.salaryDetail}\n\nIf your expectation is below market, prepare a case with 2–3 quantified wins and the skills employers list most often. If you're above market, emphasise rare skills and leadership impact.`
        : 'Salary signals are limited for your profile currency/role mix. Broaden desired roles or skills so the hub can compare you against more catalog listings.';
      break;
    case 'promotion':
      body = [
        `Readiness score: ${input.context.promotion.score}% (${input.context.promotion.level}) toward ${input.context.promotion.targetSeniority}.`,
        input.context.promotion.detail,
        input.context.promotion.checklistOpen.length
          ? `Open checklist items: ${input.context.promotion.checklistOpen.join('; ')}.`
          : 'Your promotion checklist looks largely complete — focus on evidence in interviews and applications.',
      ].join('\n\n');
      break;
    case 'career_path':
      body = input.context.careerPaths.length
        ? input.context.careerPaths
            .map(
              (p) =>
                `• ${p.title} — ${p.readinessPct}% ready. ${p.detail}`,
            )
            .join('\n')
        : 'No ladder suggestions yet. Set desired roles on your profile (e.g. Senior Backend Engineer) and reopen Growth.';
      break;
    case 'roadmap':
      body = input.context.roadmapSteps.length
        ? [
            'Suggested learning sequence from your Growth Hub:',
            ...input.context.roadmapSteps.map(
              (s, i) =>
                `${i + 1}. ${s.title}${s.estimatedHours ? ` (~${s.estimatedHours}h)` : ''}`,
            ),
            input.context.certifications.length
              ? `Optional credentials: ${input.context.certifications.map((c) => c.name).join(', ')}.`
              : '',
          ]
            .filter(Boolean)
            .join('\n')
        : 'No roadmap steps yet — fill skill gaps on your profile or wait for more active jobs in the catalog.';
      break;
    case 'skill_gaps':
    default: {
      if (input.context.topGaps.length === 0) {
        body =
          'You have strong coverage against current catalog demand. Stretch by targeting a harder seniority band or a neighbouring stack.';
      } else {
        const lines = input.context.topGaps.slice(0, 4).map(
          (g) => `• ${g.skill} (${g.priority}) — ${g.reason}`,
        );
        body = `Prioritise these gaps before polishing secondary skills:\n${lines.join('\n')}`;
      }
      break;
    }
  }

  const followUp = nextActions(input.context, inferred);
  const turnHint =
    input.history.filter((m) => m.role === 'user').length >= 2
      ? '\n\nYou can also open Interview or Practice from the Career nav to rehearse what you learn here.'
      : '';

  return {
    source: 'template',
    content: `${body}\n\n${followUp}${turnHint}`,
  };
}

function inferFocusFromMessage(text: string, fallback: CoachFocus): CoachFocus {
  if (/\b(salary|comp|pay|compensation|raise)\b/.test(text)) return 'salary';
  if (/\b(promot\w*|senior|lead|staff|level.?up)\b/.test(text)) return 'promotion';
  if (/\b(path|ladder|career.?change|next.?role)\b/.test(text)) return 'career_path';
  if (/\b(roadmap|learn|course|study|curriculum)\b/.test(text)) return 'roadmap';
  if (/\b(skill|gap|missing|stack)\b/.test(text)) return 'skill_gaps';
  return fallback;
}

function nextActions(context: CoachContextDto, focus: CoachFocus): string {
  switch (focus) {
    case 'salary':
      return 'Ask me: “How should I negotiate?” or “Which skills lift my band?”';
    case 'promotion':
      return 'Ask me: “What evidence should I collect?” or “Which skill unlocks the next level?”';
    case 'career_path':
      return 'Ask me: “Which path fits me best?” or “What should I learn for that path?”';
    case 'roadmap':
      return 'Ask me: “What should I study this week?” or “Which cert is worth it?”';
    case 'skill_gaps':
      return context.topGaps[0]
        ? `Ask me: “How do I learn ${context.topGaps[0].skill}?” or “What should I prioritise first?”`
        : 'Ask me anything about your growth plan.';
    default:
      return 'Try: skill gaps, learning roadmap, salary, promotion readiness, or career paths.';
  }
}

export function toCareerCoachSessionDto(row: {
  id: string;
  userId: string;
  status: string;
  focus: string;
  title: string | null;
  messagesJson: unknown;
  contextJson: unknown;
  summary: string | null;
  source: string;
  createdAt: Date;
  updatedAt: Date;
}): CareerCoachSessionDto {
  const messages = Array.isArray(row.messagesJson)
    ? (row.messagesJson as CoachMessageDto[])
    : [];
  const context =
    row.contextJson && typeof row.contextJson === 'object'
      ? (row.contextJson as CoachContextDto)
      : null;

  return {
    id: row.id,
    userId: row.userId,
    status: row.status,
    focus: row.focus,
    title: row.title,
    messages,
    context,
    summary: row.summary,
    source: row.source,
    createdAt: row.createdAt.toISOString(),
    updatedAt: row.updatedAt.toISOString(),
  };
}

File: apps/ai-service/app/services/coach.py
"""Career coach chat — LiteLLM with Growth Hub template fallback."""

from __future__ import annotations

import json
import logging
import time
from typing import Any

from pydantic import BaseModel, Field

from app.config import settings

logger = logging.getLogger(__name__)

MAX_MESSAGE_CHARS = 4_000
MAX_HISTORY = 12


class CoachGap(BaseModel):
    skill: str = ""
    priority: str = ""
    reason: str = ""


class CoachRoadmapStep(BaseModel):
    title: str = ""
    skill: str = ""
    estimatedHours: int | None = None


class CoachPath(BaseModel):
    title: str = ""
    readinessPct: int = 0
    detail: str = ""


class CoachPromotion(BaseModel):
    score: int = 0
    level: str = ""
    targetSeniority: str = ""
    detail: str = ""
    checklistOpen: list[str] = Field(default_factory=list)


class CoachMarket(BaseModel):
    activeJobs: int = 0
    skillsAnalyzed: int = 0


class CoachContext(BaseModel):
    summary: str = ""
    topGaps: list[CoachGap] = Field(default_factory=list)
    roadmapSteps: list[CoachRoadmapStep] = Field(default_factory=list)
    certifications: list[dict[str, Any]] = Field(default_factory=list)
    careerPaths: list[CoachPath] = Field(default_factory=list)
    salaryDetail: str | None = None
    promotion: CoachPromotion = Field(default_factory=CoachPromotion)
    market: CoachMarket = Field(default_factory=CoachMarket)
    memorySummary: str = ""
    memoryFacts: list[str] = Field(default_factory=list)


class CoachHistoryMessage(BaseModel):
    role: str
    content: str


class CoachChatInput(BaseModel):
    focus: str = "general"
    user_message: str
    context: CoachContext = Field(default_factory=CoachContext)
    messages: list[CoachHistoryMessage] = Field(default_factory=list)


class LlmMetadata(BaseModel):
    enabled: bool = False
    used: bool = False
    model: str | None = None
    error: str | None = None
    durationMs: int | None = None


class CoachChatResponse(BaseModel):
    reply: str
    source: str = "template"
    llm: LlmMetadata = Field(default_factory=LlmMetadata)


def chat(payload: CoachChatInput) -> CoachChatResponse:
    user_message = " ".join(payload.user_message.split()).strip()[:MAX_MESSAGE_CHARS]
    if not user_message:
        return CoachChatResponse(
            reply="Ask me about skill gaps, your learning roadmap, salary, promotion readiness, or career paths.",
            source="template",
        )

    template = _template_reply(payload, user_message)
    llm_meta, llm_reply = _generate_with_llm(payload, user_message)

    if llm_reply and llm_reply.strip():
        return CoachChatResponse(reply=llm_reply.strip(), source="llm", llm=llm_meta)

    return CoachChatResponse(reply=template, source="template", llm=llm_meta)


def iter_chat_events(payload: CoachChatInput):
    """Yield SSE-ready dict events: meta → token* → done (or error)."""
    user_message = " ".join(payload.user_message.split()).strip()[:MAX_MESSAGE_CHARS]
    if not user_message:
        reply = (
            "Ask me about skill gaps, your learning roadmap, salary, "
            "promotion readiness, or career paths."
        )
        yield {"type": "meta", "source": "template"}
        yield {"type": "token", "text": reply}
        yield {"type": "done", "source": "template", "reply": reply}
        return

    template = _template_reply(payload, user_message)
    model = settings.litellm_model
    meta = LlmMetadata(enabled=True, model=model)
    started = time.perf_counter()

    try:
        import litellm
    except ImportError:
        meta.enabled = False
        meta.error = "litellm is not installed (run: INSTALL_LLM=1 pnpm setup:ai)"
        yield from _yield_template(template, meta)
        return

    assembled: list[str] = []
    try:
        stream = _stream_complete(litellm, model, payload, user_message)
        first = next(stream, None)
        if first is None:
            raise RuntimeError("empty llm stream")

        yield {"type": "meta", "source": "llm", "llm": meta.model_dump()}
        assembled.append(first)
        yield {"type": "token", "text": first}
        for piece in stream:
            assembled.append(piece)
            yield {"type": "token", "text": piece}

        reply = "".join(assembled).strip()
        if not reply:
            raise RuntimeError("empty llm stream")
        meta.used = True
        meta.durationMs = int((time.perf_counter() - started) * 1000)
        yield {"type": "done", "source": "llm", "reply": reply, "llm": meta.model_dump()}
    except Exception as exc:  # noqa: BLE001
        logger.warning("coach llm stream failed: %s", exc)
        meta.error = str(exc)[:300]
        meta.durationMs = int((time.perf_counter() - started) * 1000)
        if assembled:
            reply = "".join(assembled).strip()
            yield {
                "type": "done",
                "source": "llm",
                "reply": reply,
                "llm": meta.model_dump(),
            }
            return
        yield from _yield_template(template, meta)


def _yield_template(template: str, meta: LlmMetadata):
    yield {"type": "meta", "source": "template", "llm": meta.model_dump()}
    for chunk in _chunk_text(template):
        yield {"type": "token", "text": chunk}
    yield {
        "type": "done",
        "source": "template",
        "reply": template,
        "llm": meta.model_dump(),
    }


def _chunk_text(text: str, size: int = 28) -> list[str]:
    if not text:
        return []
    return [text[i : i + size] for i in range(0, len(text), size)]


def _stream_complete(
    litellm: Any,
    model: str,
    payload: CoachChatInput,
    user_message: str,
):
    context_blob = json.dumps(payload.context.model_dump(), ensure_ascii=True)[:6_000]
    history = payload.messages[-MAX_HISTORY:]
    history_lines = "\n".join(f"{m.role}: {m.content[:800]}" for m in history)

    system = (
        "You are a concise career coach for AI JobMatch Copilot. "
        "Ground every recommendation in the provided Growth Hub context. "
        "Honor durable memorySummary/memoryFacts when present. "
        "Do not invent employers, salaries, or credentials not present in context. "
        "Keep replies under 220 words. Use short paragraphs or simple bullets. "
        "You may use light Markdown only (**bold** for short labels, - for lists). "
        "Do not use headings (#), tables, or HTML."
    )
    user = (
        f"Focus area: {payload.focus}\n"
        f"Growth Hub context JSON:\n{context_blob}\n\n"
        f"Recent messages:\n{history_lines or '(none)'}\n\n"
        f"User just said: {user_message}\n\n"
        "Respond as the coach."
    )

    response = litellm.completion(
        model=model,
        messages=[
            {"role": "system", "content": system},
            {"role": "user", "content": user},
        ],
        temperature=0.4,
        stream=True,
    )
    for chunk in response:
        try:
            delta = chunk.choices[0].delta.content
        except (AttributeError, IndexError, KeyError, TypeError):
            delta = None
        if delta:
            yield delta



def _template_reply(payload: CoachChatInput, user_message: str) -> str:
    ctx = payload.context
    text = user_message.lower()
    focus = payload.focus or "general"

    if any(k in text for k in ("salary", "comp", "pay", "compensation", "raise")):
        focus = "salary"
    elif any(k in text for k in ("promot", "senior", "lead", "staff")):
        focus = "promotion"
    elif any(k in text for k in ("path", "ladder", "career change", "next role")):
        focus = "career_path"
    elif any(k in text for k in ("roadmap", "learn", "course", "study")):
        focus = "roadmap"
    elif any(k in text for k in ("skill", "gap", "missing", "stack")):
        focus = "skill_gaps"

    if focus == "salary":
        body = ctx.salaryDetail or (
            "Salary signals are limited for your current profile mix. "
            "Update desired roles or skills so Growth Hub can compare more catalog roles."
        )
    elif focus == "promotion":
        open_items = "; ".join(ctx.promotion.checklistOpen[:4]) or "checklist looks complete"
        body = (
            f"Promotion readiness is {ctx.promotion.score}% "
            f"({ctx.promotion.level}) toward {ctx.promotion.targetSeniority or 'the next level'}. "
            f"{ctx.promotion.detail} Open items: {open_items}."
        )
    elif focus == "career_path":
        if ctx.careerPaths:
            body = " ".join(
                f"{p.title} ({p.readinessPct}% ready): {p.detail}" for p in ctx.careerPaths[:3]
            )
        else:
            body = "Add desired roles on your profile so I can suggest clearer career ladders."
    elif focus == "roadmap":
        if ctx.roadmapSteps:
            steps = "; ".join(
                f"{i + 1}. {s.title}"
                + (f" (~{s.estimatedHours}h)" if s.estimatedHours else "")
                for i, s in enumerate(ctx.roadmapSteps[:5])
            )
            body = f"Suggested learning sequence: {steps}."
        else:
            body = "No roadmap steps yet — fill skill gaps or wait for more active catalog jobs."
    else:
        if ctx.topGaps:
            gaps = "; ".join(
                f"{g.skill} ({g.priority}) — {g.reason}" for g in ctx.topGaps[:4]
            )
            body = f"Prioritise these skill gaps: {gaps}."
        else:
            body = (
                ctx.summary
                or "Your profile covers most high-demand skills. Stretch into a harder seniority band."
            )

    return f"{body} Ask a follow-up if you want a weekly plan or negotiation framing."


def _generate_with_llm(
    payload: CoachChatInput,
    user_message: str,
) -> tuple[LlmMetadata, str | None]:
    model = settings.litellm_model
    meta = LlmMetadata(enabled=True, model=model)
    started = time.perf_counter()

    try:
        import litellm
    except ImportError:
        meta.enabled = False
        meta.error = "litellm is not installed (run: INSTALL_LLM=1 pnpm setup:ai)"
        return meta, None

    try:
        reply = _complete(litellm, model, payload, user_message)
        meta.used = True
        meta.durationMs = int((time.perf_counter() - started) * 1000)
        return meta, reply
    except Exception as exc:  # noqa: BLE001 — surface as template fallback
        logger.warning("coach llm failed: %s", exc)
        meta.error = str(exc)[:300]
        meta.durationMs = int((time.perf_counter() - started) * 1000)
        return meta, None


def _complete(litellm: Any, model: str, payload: CoachChatInput, user_message: str) -> str:
    context_blob = json.dumps(payload.context.model_dump(), ensure_ascii=True)[:6_000]
    history = payload.messages[-MAX_HISTORY:]
    history_lines = "\n".join(f"{m.role}: {m.content[:800]}" for m in history)

    system = (
        "You are a concise career coach for AI JobMatch Copilot. "
        "Ground every recommendation in the provided Growth Hub context. "
        "Honor durable memorySummary/memoryFacts when present. "
        "Do not invent employers, salaries, or credentials not present in context. "
        "Keep replies under 220 words. Use short paragraphs or simple bullets. "
        "You may use light Markdown only (**bold** for short labels, - for lists). "
        "Do not use headings (#), tables, or HTML."
    )
    user = (
        f"Focus area: {payload.focus}\n"
        f"Growth Hub context JSON:\n{context_blob}\n\n"
        f"Recent messages:\n{history_lines or '(none)'}\n\n"
        f"User just said: {user_message}\n\n"
        "Respond as the coach."
    )

    kwargs: dict[str, Any] = {
        "model": model,
        "messages": [
            {"role": "system", "content": system},
            {"role": "user", "content": user},
        ],
        "temperature": 0.4,
    }
    response = litellm.completion(**kwargs)
    content = response.choices[0].message.content or ""
    return content.strip()
